File: apps/driver-ios/Packages/Core/Sources/Core/StateMachine/JobStateMachine.swift
import Foundation

/// Mirror of `apps/api/src/modules/jobs/job-state-machine.ts`.
///
/// Transition map:
///   new          → dispatched, cancelled
///   dispatched   → enroute, new, cancelled, goa
///   enroute      → on_scene, cancelled, goa
///   on_scene     → in_progress, goa, cancelled
///   in_progress  → completed, cancelled
///   completed    → (terminal)
///   cancelled    → (terminal)
///   goa          → (terminal)
public enum JobStateMachine {
    private static let transitionMap: [JobStatus: [JobStatus]] = [
        .new: [.dispatched, .cancelled],
        .dispatched: [.enroute, .new, .cancelled, .goa],
        .enroute: [.onScene, .cancelled, .goa],
        .onScene: [.inProgress, .goa, .cancelled],
        .inProgress: [.completed, .cancelled],
        .completed: [],
        .cancelled: [],
        .goa: [],
    ]

    public static let terminalStatuses: Set<JobStatus> = [.completed, .cancelled, .goa]

    public static func canTransition(from: JobStatus, to: JobStatus) -> Bool {
        transitionMap[from]?.contains(to) ?? false
    }

    public static func allowedTransitions(from: JobStatus) -> [JobStatus] {
        transitionMap[from] ?? []
    }

    public static func isTerminal(_ status: JobStatus) -> Bool {
        terminalStatuses.contains(status)
    }

    /// The primary driver-facing forward path used by the active job screen.
    public static func nextForwardStep(from: JobStatus) -> JobStatus? {
        switch from {
        case .new: return .dispatched
        case .dispatched: return .enroute
        case .enroute: return .onScene
        case .onScene: return .inProgress
        case .inProgress: return .completed
        default: return nil
        }
    }

    public static func driverActionLabel(currentStatus: JobStatus) -> String {
        switch currentStatus {
        case .new, .dispatched: return "Start En Route"
        case .enroute: return "Arrived On Scene"
        case .onScene: return "Load & Tow"
        case .inProgress: return "Complete Drop"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        case .goa: return "Gone On Arrival"
        }
    }
}

public struct InvalidJobTransitionError: Error, Equatable {
    public let from: JobStatus
    public let to: JobStatus
}

File: apps/driver-android/app/src/main/java/ai/bluecollar/ustowdispatch/driver/data/jobs/JobStateMachine.kt
package ai.bluecollar.ustowdispatch.driver.data.jobs

/**
 * Client-side mirror of apps/api/src/modules/jobs/job-state-machine.ts.
 *
 *   new          → dispatched, cancelled
 *   dispatched   → enroute, new (unassign), cancelled, goa
 *   enroute      → on_scene, cancelled, goa
 *   on_scene     → in_progress, goa, cancelled
 *   in_progress  → completed, cancelled
 *   completed    → (terminal)
 *   cancelled    → (terminal)
 *   goa          → (terminal)
 *
 * The driver UI offers only the forward path plus terminal off-ramps.
 * The "unassign" branch (dispatched → new) is dispatcher-side and never
 * surfaced in this app.
 *
 * Mirror of the iOS Core/StateMachine/JobStateMachine.swift module.
 */
object JobStatus {
    const val NEW = "new"
    const val DISPATCHED = "dispatched"
    const val ENROUTE = "enroute"
    const val ON_SCENE = "on_scene"
    const val IN_PROGRESS = "in_progress"
    const val COMPLETED = "completed"
    const val CANCELLED = "cancelled"
    const val GOA = "goa"

    val ALL: Set<String> = setOf(NEW, DISPATCHED, ENROUTE, ON_SCENE, IN_PROGRESS, COMPLETED, CANCELLED, GOA)
    val TERMINAL: Set<String> = setOf(COMPLETED, CANCELLED, GOA)
}

object JobStateMachine {
    private val TRANSITIONS: Map<String, List<String>> = mapOf(
        JobStatus.NEW to listOf(JobStatus.DISPATCHED, JobStatus.CANCELLED),
        JobStatus.DISPATCHED to listOf(JobStatus.ENROUTE, JobStatus.NEW, JobStatus.CANCELLED, JobStatus.GOA),
        JobStatus.ENROUTE to listOf(JobStatus.ON_SCENE, JobStatus.CANCELLED, JobStatus.GOA),
        JobStatus.ON_SCENE to listOf(JobStatus.IN_PROGRESS, JobStatus.GOA, JobStatus.CANCELLED),
        JobStatus.IN_PROGRESS to listOf(JobStatus.COMPLETED, JobStatus.CANCELLED),
        JobStatus.COMPLETED to emptyList(),
        JobStatus.CANCELLED to emptyList(),
        JobStatus.GOA to emptyList(),
    )

    fun canTransition(from: String, to: String): Boolean =
        TRANSITIONS[from]?.contains(to) == true

    fun allowedTransitions(from: String): List<String> = TRANSITIONS[from].orEmpty()

    fun isTerminal(status: String): Boolean = status in JobStatus.TERMINAL

    fun driverActions(from: String): List<String> = when (from) {
        JobStatus.DISPATCHED -> listOf(JobStatus.ENROUTE, JobStatus.CANCELLED, JobStatus.GOA)
        else -> allowedTransitions(from)
    }
}

class InvalidJobTransitionException(val from: String, val to: String) :
    IllegalStateException("Cannot transition job from '$from' to '$to'")
